x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/arithmetic/Neg.java

```diff
@@ -67,29 +68,30 @@ else if (type == DataTypes.LONG) {
 
     @Override
     public final Object fold() {
-        if (isTemporalAmount(field().dataType()) && field() instanceof Literal literal) {
-            return foldTemporalAmount(literal);
-        }
-        return EvaluatorMapper.super.fold();
-    }
-
-    private Object foldTemporalAmount(Literal literal) {
-        try {
-            Object value = literal.fold();
-            if (value instanceof Period period) {
-                return period.negated();
+        DataType dataType = field().dataType();
+        // For date periods and time durations, we need to handle folding especially. These types are unrepresentable, so there is no
+        // evaluator for them - but the default folding requires an evaluator.
+        if (dataType == DATE_PERIOD) {
+            Period fieldValue = (Period) field().fold();
+            try {
```

Comment:
nit: maybe extract the try-catch into a method, to lower duplication?

Reply:
Tried that, annoyingly doesn't really seem to be possible.

`Period::negated` and `Duration::negated` are not implementations of a common interface, so I cannot abstract that away and need to branch based on the data type :/